src/v/kafka/server/connection_context.cc

```diff
@@ -353,169 +355,143 @@ connection_context::reserve_request_units(api_key key, size_t size) {
-                              [self, sres](std::exception_ptr e) {
-                                  // ssx::spawn_with_gate already caught
-                                  // shutdown-like exceptions, so we should only
-                                  // be taking this path for real errors.  That
-                                  // also means that on shutdown we don't bother
-                                  // to call shutdown_input on the connection,
-                                  // so rely on any future reader to check the
-                                  // abort source before considering reading the
-                                  // connection.
-                                  auto disconnected
-                                    = net::is_disconnect_exception(e);
-                                  if (disconnected) {
-                                      vlog(
-                                        klog.info,
-                                        "Disconnected {} ({})",
-                                        self->conn->addr,
-                                        disconnected.value());
-                                  } else {
-                                      vlog(
-                                        klog.warn,
-                                        "Error processing request: {}",
-                                        e);
-                                  }
-
-                                  sres->tracker->mark_errored();
-                                  self->conn->shutdown_input();
-                              });
-                      return d;
-                  })
-                  .handle_exception([self, sres](std::exception_ptr e) {
-                      vlog(
-                        klog.info, "Detected error dispatching request: {}", e);
-                      self->conn->shutdown_input();
-                      sres->tracker->mark_errored();
-                  });
-            });
-      })
-      .handle_exception_type([](const ss::sleep_aborted&) {
-          // shutdown started while force-throttling
-          return ss::make_ready_future<>();
-      });
+    auto sres_in = co_await throttle_request(hdr, size);
+    if (_server.abort_requested()) {
+        // protect against shutdown behavior
+        co_return;
+    }
+    if (_kafka_throughput_controlled_api_keys().at(hdr.key)) {
+        // Normally we can only get here after a prior call to
+        // snc_quota_mgr().get_or_create_quota_context() in
+        // record_tp_and_calculate_throttle(), but there is possibility
+        // that the changing configuration could still take us into this
+        // branch with unmatching (and even null) _snc_quota_context.
+        // Simply an unmatching _snc_quota_context is no big deal because
+        // it is a one off event, but we need protection from it being
+        // nullptr
+        if (likely(_snc_quota_context)) {
```

Comment:
While we're talking about cleanups is it time to switch to c++20 attributes `[[likely]]` and `[[unlikely]]` instead of these macros?

Reference: https://en.cppreference.com/w/cpp/language/attributes/likely

Reply:
Yeh good call. I think at one point we had c++20 but the compilers hadn't added this yet but now I see at least one usage of these in the tree.